File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-3.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        #sorted

        if nums[0] <= nums[-1] :
            return nums[0]
        
        #not sorted
        l=0
        r = len(nums) - 1

        while(l<=r):
            mid = (l+r)//2
            if(nums[mid+1] > nums[mid] and nums[mid] > nums[0]):
                l = mid
            elif(nums[mid+1] > nums[mid] and nums[mid] < nums[0]):
                r = mid
            else:
                return nums[mid+1]
        
        return -1

File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> nums;

    void dfs(TreeNode* root){
        if(!root){
            return;
        }

        dfs(root->left);
        nums.push_back(root->val);
        dfs(root->right);
    }
    bool isValidBST(TreeNode* root) {
        dfs(root);
        for(int i=0;i<nums.size()-1;i++){
            if(nums[i]>=nums[i+1]){
                return false;
            }
        }
        return true;
    }
};
